src/ast/codegen_llvm.cpp

```diff
@@ -994,8 +994,15 @@ void CodegenLLVM::visit(Unop &unop)
     {
       case bpftrace::Parser::token::LNOT:
       {
-        Value *zero_value = Constant::getNullValue(expr_->getType());
+        auto ty = expr_->getType();
+        Value *zero_value = Constant::getNullValue(ty);
         expr_ = b_.CreateICmpEQ(expr_, zero_value);
+        // CreateICmpEQ() returns 1-bit integer
+        // Cast it to the same type of the operand
+        // Use unsigned extention first, otherwise !0 becomes -1
+        expr_ = b_.CreateIntCast(expr_, ty, false);
+        if (unop.expr->type.is_signed)
+          expr_ = b_.CreateIntCast(expr_, ty, true);
```

Comment:
I realized that LLVM does not distinguish signed and unsigned, so this is not needed.